pkg/tests/testsupport/metric_loader.go

```diff
@@ -200,29 +232,40 @@ func (si *sampleIngestor) shardSamples() {
 			si.shards[shardIdx] <- sample.Val
 			if si.rate != nil {
 				if err := si.rate.Wait(context.Background()); err != nil {
-					log.Error(err)
+					log.Error("msg", err)
 				}
 			}
 		}
 	}()
 }
 
 func (si *sampleIngestor) ingestSamples(ingest IngestFunc) {
-	var wg sync.WaitGroup
+	var shardWg sync.WaitGroup
+	var ingestWg sync.WaitGroup
+	reqCh := make(chan prompb.WriteRequest, 100)
```

Comment:
So I'm not sure why the shards are in separate goroutines... IMO this channel should be the size of the number of shards and you spin up the ingestor goroutines and then load shards into the channel sequentially. Close the channel to stop the ingestors and then you need only one waitgroup to wait on.

Reply:
We need parallel workers on `reqCh` because otherwise things will run sequentially meaning that one shard will block another. Also we want to prepare batches for each shard in their own routines. I hope this makes it more clear or am I missing something?